File: Java-CodeServer/src/main/resources/js/prompto/type/SetType.js
var CollectionType = require("./CollectionType").CollectionType;
var IntegerType = require("./IntegerType").IntegerType;
var BooleanType = require("./BooleanType").BooleanType;

function SetType(itemType) {
	CollectionType.call(this, itemType.name+"<>", itemType);
	this.itemType = itemType;
	return this;
}

SetType.prototype = Object.create(CollectionType.prototype);
SetType.prototype.constructor = SetType;


SetType.prototype.equals = function(obj) {
	if(obj===this) {
		return true;
	}
	if(obj===null) {
		return false;
	}
	if(!(obj instanceof SetType)) {
		return false;
	}
	return this.itemType.equals(obj.itemType);
};

SetType.prototype.checkAdd = function(context, other, tryReverse) {
	if(other instanceof SetType && this.itemType.equals(other.itemType, tryReverse)) {
		return this;
	} else {
		return CollectionType.prototype.checkAdd.call(this, context, other);
	}
};

SetType.prototype.checkItem = function(context, other) {
	if(other==IntegerType.instance) {
		return this.itemType;
	} else {
		return CollectionType.prototype.checkItem.call(this, context, other);
	}
};

SetType.prototype.checkContainsAllOrAny = function(context, other) {
	return BooleanType.instance;
}

SetType.prototype.checkIterator = function(context) {
	return this.itemType;
}

SetType.prototype.checkMember = function(context, name) {
	if ("length" == name) {
		return IntegerType.instance;
	} else {
		return CollectionType.prototype.checkMember.call(this, context, name);
	}
};

exports.SetType = SetType;

File: Java-Core/src/main/java/prompto/type/SetType.java
package prompto.type;

import java.lang.reflect.Type;

import prompto.grammar.Identifier;
import prompto.intrinsic.PromptoSet;
import prompto.runtime.Context;


public class SetType extends ContainerType {

	public SetType(IType itemType) {
		super(Family.SET, itemType, itemType.getTypeName()+"<>");
	}
	
	@Override
	public Type getJavaType() {
		return PromptoSet.class;
	}
	
	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return (other instanceof SetType) && itemType.isAssignableTo(context, ((SetType)other).getItemType());
	}

	@Override
	public IType checkIterator(Context context) {
		return itemType;
	}
	
	@Override
	public IType checkMember(Context context, Identifier id) {
		String name = id.toString();
        if ("length".equals(name))
            return IntegerType.instance();
        else
    		return super.checkMember(context, id);
    }

	@Override
	public IType checkAdd(Context context, IType other, boolean tryReverse) {
		if(other instanceof ContainerType) {
			IType itemType = ((ContainerType)other).getItemType();
			if((other instanceof ListType || other instanceof SetType) 
					&& this.getItemType().equals(itemType))
				return this;
		} 
		return super.checkAdd(context, other, tryReverse);
	}
	
	@Override
	public IType checkItem(Context context, IType other) {
		if(other==IntegerType.instance())
			return itemType;
		else
			return super.checkItem(context,other);
	}

	@Override
	public IType checkContainsAllOrAny(Context context, IType other) {
		return BooleanType.instance();
	}
	

}
